lib/Dialect/FIRRTL/FIRRTLOpInterfaces.cpp

```diff
@@ -20,6 +20,63 @@ using namespace mlir;
 using namespace llvm;
 using namespace circt::firrtl;
 
+//===----------------------------------------------------------------------===//
+// FlowKind
+//===----------------------------------------------------------------------===//
+
+Flow circt::firrtl::flipFlow(Flow flow) {
+  switch (flow) {
+  case Flow::Source:
+    return Flow::Sink;
+  case Flow::Sink:
+    return Flow::Source;
+  case Flow::Duplex:
+    return Flow::Duplex;
+  }
+}
+
+static Flow mergeFlow(Flow a, Flow b) {
+  switch (a) {
+  case Flow::Source:
+    return b;
+  case Flow::Sink:
+    return flipFlow(b);
+  case Flow::Duplex:
+    return Flow::Duplex;
+  }
+}
+
+Flow FlowKind::get(Value value) {
+  Flow flow = Flow::Source;
+
+  do {
+    if (auto blockArg = value.dyn_cast<BlockArgument>()) {
+      auto op = dyn_cast<FlowKind>(value.getParentBlock()->getParentOp());
+      // If the operation doesn't implement the interface, return the result
+      // so far.
+      if (!op)
+        return flow;
+      auto result = op.getBlockArgFlow(blockArg);
+      flow = mergeFlow(result.getFlow(), flow);
+      value = result.getValue();
+    } else {
+      auto op = dyn_cast<FlowKind>(value.getDefiningOp());
+      // If the operation doesn't implement the interface, return the result
+      // so far.
+      if (!op)
+        return flow;
+      auto result = op.getFlow(value.cast<OpResult>());
+      flow = mergeFlow(result.getFlow(), flow);
+      value = result.getValue();
+    }
```

Comment:
You may be able to share some code here by defining a `FlowResult result;` outside the if/else, setting that inside, and then keeping `flow = ...;' and `value = ...;' common after the if/else.